// appbuilder/lib/nativescript-resources.js
"use strict";
const path = require("path");
class NativeScriptResources {
    constructor($projectConstants, $resources) {
        this.$projectConstants = $projectConstants;
        this.$resources = $resources;
    }
    get nativeScriptResourcesDir() {
        if (!this._nativeScriptResourcesDir) {
            this._nativeScriptResourcesDir = this.$resources.resolvePath("NativeScript");
        }
        return this._nativeScriptResourcesDir;
    }
    get nativeScriptMigrationFile() {
        if (!this._nativeScriptMigrationFile) {
            this._nativeScriptMigrationFile = path.join(this.nativeScriptResourcesDir, "nativeScript-migration-data.json");
        }
        return this._nativeScriptMigrationFile;
    }
    get nativeScriptDefaultPackageJsonFile() {
        if (!this._nativeScriptDefaultPackageJsonFile) {
            this._nativeScriptDefaultPackageJsonFile = path.join(this.nativeScriptResourcesDir, this.$projectConstants.PACKAGE_JSON_NAME);
        }
        return this._nativeScriptDefaultPackageJsonFile;
    }
}
exports.NativeScriptResources = NativeScriptResources;
$injector.register("nativeScriptResources", NativeScriptResources);
